<a href="https://colab.research.google.com/github/pranavpatnaik1/Deep-Learning-Projects-Using-TensorFlow-2/blob/main/ch2/digit_recognition_mnist.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Training a Model for Digit Recognition: MNIST

Here, we'll cover how we can train a model via MNIST to perform digit recognition. It's fairly simple and is rather inexpensive.

Importing modules: Keras, Tensorflow

In [1]:
import keras
import tensorflow as tf

Next, we'll read in the data. Since Keras comes pre-loaded with MNIST, we'll use that.

In [3]:
mnist=tf.keras.datasets.mnist

# split data
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# flattening data
x_train, x_test = x_train / 225.0, x_test / 255.0

Moving on, we're creating the neural network via the *Sequential* module. The loss function will be Sparse Categorical Cross Entropy (for 10 options: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9). Moreover, we'll use the Adam optimizer.

In [4]:
digit = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28)),
    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

# compiling model
digit.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

/usr/local/lib/python3.10/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Training the model:

In [5]:
digit.fit(x_train, y_train, epochs=3)

Epoch 1/3
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - accuracy: 0.8171 - loss: 0.6804
Epoch 2/3
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - accuracy: 0.9169 - loss: 0.2999
Epoch 3/3
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9196 - loss: 0.2815


Finally, let's evaluate the model

In [6]:
digit.evaluate(x_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9141 - loss: 0.3182


[0.2811993360519409, 0.9239000082015991]

With simply 3 epochs (3 full runthroughs), we've achieved an accuracy score of 92%. Pretty great, yeah?

That's it for digit recognition via MNIST. Hope that helps!


-Pranav